RFC-0026-logging-system.md

```diff
@@ -0,0 +1,481 @@
+info classes do not currently exist. Also, only one type of warning currently
+exists in C++, and it is not implemented as a C++ class that can be inherited
+(as far as I understand).
+
+#### Error message classes:
+  
+* **`c10::Error`** - Python `RuntimeError`
+  - Default error class. Other error classes inherit from it.
+
+* **`c10::IndexError`** - Python `IndexError`
+  - Emitted when attempting to access an element that is not present in
+    a list-like object.
+
+* **`c10::ValueError`** - Python `ValueError`
+  - Emitted when a function receives an argument with correct type but
+    incorrect value.
+
+* **`c10::TypeError`** - Python `TypeError`
+  - Emitted when a function receives an argument with incorrect type.
+
+* **`c10:NotImplementedError`** - Python `NotImplementedError`
+  - Emitted when a feature that is not implemented is called.
+
+* **`c10::LinAlgError`** - Python `torch.linalg.LinAlgError`
+  - Emitted from the `torch.linalg` module when there is a numerical error.
+
+* **`c10::NondeterministicError`** - Python `torch.NondeterministicError`
+  - Emitted when `torch.use_deterministic_algorithms(True)` and
+    `torch.set_deterministic_debug_mode('error')` are set, and a
+    nondeterministic operation is called.
+  
+
+#### Warning message classes:
+
+* **`c10::UserWarning`** - Python `UserWarning`
+  - Default warning class. Other warning classes inherit from it.
+
+* **`c10::BetaWarning`** - Python `torch.BetaWarning`
+  - Emitted when a beta feature is called. See
+    [PyTorch feature classifications](https://pytorch.org/blog/pytorch-feature-classification-changes/).
+
+* **`c10::PrototypeWarning`** - Python `torch.PrototypeWarning`
+  - Emitted when a prototype feature is called. See
+    [PyTorch feature classifications](https://pytorch.org/blog/pytorch-feature-classification-changes/).
+
+* **`c10::NondeterministicWarning`** - Python `torch.NondeterministicWarning`
+  - Emitted when `torch.use_deterministic_algorithms(True)` and
+    `torch.set_deterministic_debug_mode('warn')` are set, and a
+    nondeterministic operation is called.
+
+* **`c10::DeprecationWarning`** - Python `DeprecationWarning`
+  - Emitted when a deprecated function is called.
+  - TODO: `DeprecationWarning`s are ignored by default in Python, so we may
+    actually want to use a different Python class for this.
+
+
+#### Info message classes:
+  
+* **`c10::Info`** - Python `torch.Info`
+  - Default info class. Other info classes inherit from it.
+
+
+### Message APIs
+
+In order to emit messages, developers can use the APIs defined in this section.
+
+These APIs all have a variable length argument list, `...` in C++ and `*args`
+in Python. When a message is emitted, these arguments are concatenated into
+a string, and the string becomes the body of the message. In C++, the arguments
+must all have the `std::ostream& operator<<` function defined so that they can
+be concatenated, and in Python, they must all have a `__str__` function.
```

Comment:
Right, in C++, `very_expensive_function()` won't be called if `cond == true`. Same for `WARN_ONCE(very_expensive_function())`, if it already has warned once before. I'll add that detail to the doc.

I might as well ask, even though I'm almost sure the answer is no--is there any possible way to avoid calling `very_expensive_function()` in Python as well for `torch.check(true, very_expensive_function())`?

Reply:
> I might as well ask, even though I'm almost sure the answer is no--is there any possible way to avoid calling very_expensive_function() in Python as well for torch.check(true, very_expensive_function())?

I don't think so indeed. Maybe we want to consider that this function takes a callable as the second argument to preserve this property?
Or maybe this is ok and we will rely on Dynamo to get this behavior cc @ezyang ?

Reply:
Just pass it in a lambda. This is how torch._prims_common.check works